misc-scripts/xref_mapping/XrefParser/UniProtParser.pm

```diff
@@ -17,521 +17,163 @@ limitations under the License.
-    my $xref;
+use parent qw( XrefParser::BaseParser );
 
-    # set accession (and synonyms if more than one)
-    # AC line may have primary accession and possibly several ; separated synonyms
-    # May also be more than one AC line
-    my ($acc) = $_ =~ /(\nAC\s+.+)/s; # will match first AC line and everything else
 
-    my @all_lines = split /\n/, $acc;
+Readonly my $DEFAULT_LOADER_BATCH_SIZE => 1000;
 
-    # Check for CC (caution) lines containing certain text
-    # If sequence is from Ensembl, do not use
-    my $ensembl_derived_protein = 0;
-    if ($_ =~ /CAUTION: The sequence shown here is derived from an Ensembl/) {
-      $ensembl_derived_protein = 1;
-      $ensembl_derived_protein_count++;
-    }
+Readonly my %source_name_for_section => (
+                                         'Swiss-Prot' => 'Uniprot/SWISSPROT',
+                                         'TrEMBL'     => 'Uniprot/SPTREMBL',
+                                       );
 
-    # extract ^AC lines only & build list of accessions
-    my @accessions;
-    foreach my $line (@all_lines) {
-      my ($accessions_only) = $line =~ /^AC\s+(.+)/;
-      push(@accessions, (split /;\s*/, $accessions_only)) if ($accessions_only);
-
-    }
-
-
-    if(lc($accessions[0]) eq "unreviewed"){
-      print "WARNING: entries with accession of $acc not allowed will be skipped\n";
-      next;
-    }
-    $xref->{INFO_TYPE} = "SEQUENCE_MATCH";
-    $xref->{ACCESSION} = $accessions[0];
-    for (my $a=1; $a <= $#accessions; $a++) {
-      push(@{$xref->{"SYNONYMS"} }, $accessions[$a]);
-    }
 
-    my ($label, $sp_type) = $_ =~ /ID\s+(\w+)\s+(\w+)/;
-    my ($protein_evidence_code) = $_ =~ /PE\s+(\d+)/; 
-
-    # SwissProt/SPTrEMBL are differentiated by having STANDARD/PRELIMINARY here
-    if ($sp_type =~ /^Reviewed/i) {
-
-      $xref->{SOURCE_ID} = $sp_source_id;
-      $num_sp++;
-    } elsif ($sp_type =~ /Unreviewed/i) {
+sub run {
+  my ( $self, $ref_arg ) = @_;
+
+  my $general_source_id = $ref_arg->{source_id};
+  my $species_id        = $ref_arg->{species_id};
+  my $species_name      = $ref_arg->{species};
+  my $files             = $ref_arg->{files};
+  my $release_file      = $ref_arg->{rel_file};
+  my $verbose           = $ref_arg->{verbose} // 0;
+  my $dbh               = $ref_arg->{dbi} // $self->dbi;
+
+  my $loader_batch_size
+    = $ref_arg->{loader_batch_size} // $DEFAULT_LOADER_BATCH_SIZE;
+
+  if ( ( !defined $general_source_id ) or
+       ( !defined $species_id ) or
+       ( !defined $files ) )
+  {
+    croak "Need to pass source_id, species_id and files as pairs";
+  }
 
-    #Use normal source only if it is PE levels 1 & 2
-      if (defined($protein_evidence_code) && $protein_evidence_code < 3) {
-          $xref->{SOURCE_ID} = $sptr_source_id;
-          $num_sptr++;
-      } else {
-          $xref->{SOURCE_ID} = $sptr_non_display_source_id;
-          $num_sptr_non_display++;	  
+  my $extractor = XrefParser::UniProtParser::Extractor->new({
```

Comment:
@mkszuba @nerdstrike I've just realised that I won't be able to fix `WormbaseCElegansUniProtParser` if you do this unless I mess around with %INC to substitute my own `XrefParser::UniProtParser::Transformer` or something like that.

I think separating E, T, and L is a good idea into logical blocks and it's good that you've identified breaks in the previous monolith that separate it into more reasonable chunks.

I don't like where you've drawn a division between E and T. In other parsers (e.g. Tiago's new RefSeqGPFFParser) extracting means separating chunks of text at a time. Then transforming is taking the data's original format, and returning a maybe-record in an application's internal format. You do a lot of transforming in your extractor, and a transformer also filters (skipping unreviewed entries and stuff). Secondly, `XrefParser::UniProtParser::Extractor` requiring a `baseParserInstance` is also apparent enough that it's inconvenient to make it its own object.

Consider keeping extracting in the `XrefParser::UniProtParser`, where you can draw from `XrefParser::BaseParser`. @nerdstrike mentioned you'll be making a `XrefParser::BaseAdaptor` instead or something - maybe you won't even need much or any UniProt-specific behaviour to get a clean chunk of code responsible for knowing where the file is, and then opening a stream of text chunks.

I recommend you move the logic of examining the record to build Ensembl-specific things from it contained in the T part. Make it a self contained module taking a piece of text (perhaps a ref to a scalar to avoid copying) and returning either nothing - if the entry should be rejected - or an Ensembl object. It's a fantastic opportunity for encapsulation. It will be maintainable, and will let you do robust unit testing - get some canned records, say what Ensembl objects they should make, compare with `is_deeply`, and you can very easily get to 100% test coverage for it.

Finally I think your Loader should become a utility class, say a `RecordBatchingQueue`. This is the crux of your Loader really, and it's not specific to UniProt, it will work on any xrefs with dependent xrefs that it can count. Counting xrefs should not be tied to also calling the parent. The behaviour you've designed is very general and possibly so useful that it's a shame if all the other modules have to batching in their own way. Move it somewhere, think how to make it general (Java's iterators constructed from other iterators, and registering a callback that processes a full batch, are two patterns that come to my mind), add some unit tests for it, and you can solve the problem of parsed xrefs overloading the database for all the other 25981 parsers too.

If you think the current design with three objects is superior, and also you don't want to create your E, T, and L using dependency injection, could you at least put the transformer in a `sub transform`? If there is a separate sub that creates a XrefParser::UniProtParser::Transformer and applies it to the arguments, I will be able to rewrite `WormbaseCElegansUniProtParser` to inherit from it - but I won't be able to use this code in its current form.